src/libs/ajax/loaded-data/useLoadedData.ts

```diff
@@ -0,0 +1,65 @@
+import { useState } from 'react'
+import LoadedState from 'src/libs/type-utils/LoadedState'
+import { isPromiseReturn } from 'src/libs/type-utils/type-helpers'
+
+/**
+ * The Tuple returned by useLoadedData custom helper hook
+ */
+export type UseLoadedDataResult<T> = [
+  LoadedState<T, unknown>,
+  (dataCall: () => Promise<T>) => Promise<void>
+]
+
+/**
+ * A custom helper hook that will handle typical async data call mechanics and translate the possible outcomes to
+ * the appropriate LoadedState<T> result.  Initial ('None'), 'Loading', 'Error' and 'Ready' states are handled.  The
+ * Error case also handles error object as Fetch Response and extract the error message from response.text().
+ *
+ * @example
+ * const [myData, updateMyData] = useLoadedData<MyDataType>()
+ * //...
+ * updateMyData(async () => {
+ *   // any errors thrown by data call or additional checks here
+ *   // will be translated to status: 'Error' LoadedState<T>
+ *   cost coolData: MyDataType = await someDataMethod(args)
+ *   return coolData
+ * }
+ * // ...
+ * if (myData.status === 'Ready') {
+ *   const goodData = myData.state
+ *   // ...
+ * }
+ * @returns a tuple with [currentLoadedState, updateDataMethod]
+ */
+export const useLoadedData = <T>(): UseLoadedDataResult<T> => {
+  const [loadedData, setLoadedData] = useState<LoadedState<T, unknown>>({ status: 'None' })
+
+  const updateDataFn = async (dataCall: () => Promise<T>) => {
+    const previousState = loadedData.status !== 'None' ? loadedData.state : null
+    setLoadedData({
+      status: 'Loading',
+      state: previousState
+    })
+    try {
+      const result = await dataCall()
+      setLoadedData({
+        status: 'Ready',
+        state: result
+      })
+    } catch (err: unknown) {
+      const error = isPromiseReturn((err as Response).text) ?
+        Error(await (err as Response).text()) :
+        err
```

Comment:
Since `isPromiseReturn` calls the function, this won't work. The second `text` will error with "TypeError: Failed to execute 'text' on 'Response': body stream already read". Suggest checking `if (err instanceof Response) { ... }` instead.

Reply:
Not sure if maybe there's a better way to extract errors from Responses. I'd expect most error responses to contain JSON, so this would produce an error with a message containing a JSON encoded string. I guess `dataCall` could handle error responses and rethrow an error though...

Reply:
yes... this is meant to be an initial iterative solution.  I have it on my list to spend some time thinking more deeply about our error handling mechanics and related details.